### PR TITLE
dec265: support for per-image text dumping

### PR DESCRIPTION
**Pull request by chemag**

### CODE REVIEW

File: INSTALL [hunk 1]
```diff
@@ -0,0 +1,368 @@
+Installation Instructions
```
Comment (fancycode) [inline INSTALL]: This file is automatically generated by `autogen.sh` and should not be comitted to the repository.

File: libde265/de265.h [hunk 2]
```diff
@@ -378,13 +378,14 @@ enum de265_param {
...
   DE265_DECODER_PARAM_DUMP_PPS_HEADERS=3,
   DE265_DECODER_PARAM_DUMP_SLICE_HEADERS=4,
-  DE265_DECODER_PARAM_ACCELERATION_CODE=5,   // (int)  enum de265_acceleration, default: AUTO
-  DE265_DECODER_PARAM_SUPPRESS_FAULTY_PICTURES=6, // (bool)  do not output frames with decoding errors, default: no (output all images)
-
-  DE265_DECODER_PARAM_DISABLE_DEBLOCKING=7,   // (bool)  disable deblocking
-  DE265_DECODER_PARAM_DISABLE_SAO=8           // (bool)  disable SAO filter
-  //DE265_DECODER_PARAM_DISABLE_MC_RESIDUAL_IDCT=9,     // (bool)  disable decoding of IDCT residuals in MC blocks
-  //DE265_DECODER_PARAM_DISABLE_INTRA_RESIDUAL_IDCT=10  // (bool)  disable decoding of IDCT residuals in MC blocks
+  DE265_DECODER_PARAM_DUMP_IMAGE_DATA=5,
```
Comment (fancycode) [inline libde265/de265.h]: The values of existing params should not change to keep compatibility with existing applications. Just add the new property at the end and assign a previously unused value.

File: .gitignore [hunk 3]
```diff
@@ -21,3 +21,51 @@ de265-config.h
 *.ncb
 *.suo
 *.vcproj*
+
```
Comment (fancycode) [inline .gitignore]: Changes to `.gitignore` should be in a separate PR as they are unrelated to the new feature. Also you add a couple of entries that are already ignored.

**Comment by farindk** [comment]
Hello chemag,
thanks for your contribution.

I had a look at your code. The purpose of this appears to be to log a histogram of QP values used in an image. 
I assume that you are using this for some research task or to analyze some encoder behavior. However, I do not think this is of general interest enough to include it in the master branch. To me, a branch in your research repository seems a better place to maintain this.
Alternatively, we could build a framework in libde265 to plug in analyzer-modules, like yours. But that is much more work to do (and at the moment not a high priority).

BTW: I noticed that when you build the histogram, you are not weighting the QPs by CU size. Please check whether you really intend it like this, or whether this is a bug.

Dirk

File: INSTALL (hunk 1 above)
Comment (chemag) [reply INSTALL]: done

File: .gitignore (hunk 3 above)
Comment (chemag) [reply .gitignore]: done

Comment (chemag) [reply .gitignore]: https://github.com/strukturag/libde265/pull/202

File: libde265/de265.h (hunk 2 above)
Comment (chemag) [reply libde265/de265.h]: done

**Comment by chemag** [comment]
Hi Dirk,

Thanks for the quick review.

> Hello chemag,
> thanks for your contribution.
> 
> I had a look at your code. The purpose of this appears to be to log a histogram of QP values used in an image.
> I assume that you are using this for some research task or to analyze some encoder behavior. However, I do not think this is of general interest enough to include it in the master branch. To me, a branch in your research repository seems a better place to maintain this.

The idea was to have a per-image dumper (which mostly coincides with a per-slice data dumper in most streams). QP values is what I care right now, but you could also e.g. add MVs there. I've been trying to script around these two for a while, and libde265 looks like the best FOSS hevc parser around for this.

Would it make more sense to add this as a separate tool under `tools/` ? That seems to contain a hodge-podge of binaries.

> Alternatively, we could build a framework in libde265 to plug in analyzer-modules, like yours. But that is much more work to do (and at the moment not a high priority).
> 
> BTW: I noticed that when you build the histogram, you are not weighting the QPs by CU size. Please check whether you really intend it like this, or whether this is a bug.
For my use case, it does not really matter. But for a more generic, I agree it's a bug. I'll fix it :)

> 
> Dirk

**Comment by chemag** [comment]
Hmm... Trying to reopen this

**Comment by chemag** [comment]
> Would it make more sense to add this as a separate tool under `tools/` ? That seems to contain a hodge-podge of binaries.

Ping
